services/MerchStoreService.ts

```diff
@@ -279,6 +310,72 @@ export default class MerchStoreService {
     });
   }
 
+  /**
+   * TODO: add global variable
+   * Verify that items have valid options. An item with variants disabled cannot have multiple
+   * options, and an item with variants enabled cannot have multiple option types.
+   */
+  private static verifyItemHasValidPhotos(item: MerchItem | MerchandiseItemModel) {
+    if (item.photos.length > 5) {
+      throw new UserError('Merch items cannot have more than 5 pictures');
+    }
+  }
+
+  /**
+   * Creates an item photo and assign it the corresponding picture url
+   * and append the photo to the photos list from merchItem
+   * @param item merch item uuid
+   * @param properties merch item photo picture url and position
+   * @returns created item photo
+   */
+  public async createItemPhoto(item: Uuid, properties: MerchItemPhoto): Promise<PublicMerchItemPhoto> {
+    return this.transactions.readWrite(async (txn) => {
+      const merchItem = await Repositories.merchStoreItem(txn).findByUuid(item);
+      if (!merchItem) throw new NotFoundError('Merch item not found');
+
+      // add to the end
+      const position = merchItem.photos.length;
+
+      const createdPhoto = MerchandiseItemPhotoModel.create({ ...properties, position, merchItem });
+      const merchStoreItemPhotoRepository = Repositories.merchStoreItemPhoto(txn);
+
+      // verify the result photos array
+      merchItem.photos.push(createdPhoto);
+      MerchStoreService.verifyItemHasValidPhotos(merchItem);
+
+      const upsertedPhoto = await merchStoreItemPhotoRepository.upsertMerchItemPhoto(createdPhoto);
+      return upsertedPhoto.getPublicMerchItemPhoto();
+    });
+  }
+
+  /**
+   * Deletes the given item photo. Fail if the merch item is visible
+   * and it was the only photo of the item.
+   *
+   * @param uuid photo uuid
+   */
+  public async deleteItemPhoto(uuid: Uuid): Promise<void> {
+    await this.transactions.readWrite(async (txn) => {
+      const merchStoreItemPhotoRepository = Repositories.merchStoreItemPhoto(txn);
+      const photo = await merchStoreItemPhotoRepository.findByUuid(uuid);
+      if (!photo) throw new NotFoundError('Merch item photo not found');
+
+      const merchItem = await Repositories.merchStoreItem(txn).findByUuid(photo.merchItem.uuid);
+      if (merchItem.photos.length === 1 && !merchItem.hidden) {
+        throw new UserError('Cannot delete the only photo for a visible merch item');
+      }
+      // decrement photo index for all photos after the inserted photo
+      const { position } = photo;
+      merchItem.photos.forEach((p) => {
+        if (p.position > position) {
+          merchStoreItemPhotoRepository.upsertMerchItemPhoto(p, { position: p.position - 1 });
+        }
+      });
```

Comment:
Per comment above on positions, no need for this logic - let the frontend handle positions and indexes as they like.

Reply:
deleted